src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
Finding `top` doesn't involve a walk (beyond moving out of a binding pattern in the declaration), though?

Finding `top` is a matter of going from here:

```ts
function foo() {
  let {x}: {x: string | number} = {x: 1};
  //   ^
  if (typeof x === "string") {
    x;
  }
}
```
to the enclosing block.

The actual walk up starts at the node to narrow at:
```ts
function foo() {
  let {x}: {x: string | number} = {x: 1};
  if (typeof x === "string") {
    x;
 // ^
  }
}
```
and can either go up to the top level of the file/namespace (though this could traverse a lot of excess nodes if this is the only terminating condition), or can stop at the location defined by `top` - the scope within which the symbol in question is defined.